documentation/fortran-ir.md

```diff
@@ -0,0 +1,194 @@
+10. The IR will be in SSA form
+
+### Explicit Control Flow
+
+In Fortran, there are a number of statements that result in control flow to statements other than the one immediately subsequent. These can be sorted these into two categories: structured and unstructured.
+
+#### Structured Control Flow
+
+Fortran has executable constructs that imply three basic control flow forms.  The first form is a structured loop (DO construct)<sup>2</sup>. The second form is a structured cascade of conditional branches (IF construct, IF statement,<sup>3</sup> WHERE construct).  The third form is a structured multiway branch (SELECT CASE, SELECT RANK, and SELECT TYPE constructs).  The FORALL construct, while it implies a semantic model of interleaved iterations, can be modeled as a special single-entry single-exit region in FIR perhaps with backstage marker statements.<sup>4</sup>
+
+The CYCLE and EXIT statements interact with the above structured executable constructs by providing structured transfers of control.<sup>5</sup> CYCLE (possibly named) is only valid in DO constructs and creates an alternate backedge in the enclosing loop.  EXIT transfers control out of the enclosing (possibly named) construct, which need not be a DO construct.
+
+#### Unstructured Control Flow
+
+Fortran also has mechanisms of transferring control between a statement and another statement with a corresponding label.  The origin of these edges can be GOTO statements, computed GOTO statements, assigned GOTO statements, arithmetic IF statements, alt-return specifications, and END/EOR/ERR I/O specifiers.  These statements are "unstructured" in the sense that the target of the control-flow has fewer constraints and the labelled statements must be linked to their origins.
+
+Another category of unstructured control flow are statements that terminate execution.  These include RETURN, FAIL IMAGE, STOP and ERROR STOP statements.  The PAUSE statement can be modeled as a call to the runtime.
+
+### Operations
+
+The compiler's to be determined optimization passes will inform us as to the exact composition of FIR at the operations level.  This details here will necessarily change, so please read them with a grain of salt.
+
+The plan (see CFG-WP) is that statements (actions) will be a veneer model of Fortran syntactical executable constructs. Fortran statements will correspond one to one with actions. Actions will be composed of and own objects of Fortran::evaluate::GenericExprWrapper. Values of type GenericExprWrapper will have Fortran types. This implies that actions will not be in an explicit data flow representation and have optional type information.<sup>6</sup> Initially, values will bind to symbols in a context and have an implicit use-def relation. An action statement may entail a "big step" operation with many side-effects. No semantics has been defined at this time.  Actions may reference other non-executable statements from the parse tree in some to be determined manner.
```

Comment:
To clarify: all types will have been resolved by the time the CFG is created, and so expressions that appear in actions will use the least general `Expr<>` specialization.  For example, predicates in `IF` statements will all be expressions that have a concrete LOGICAL type with kind=1, not a more general expression type.  Assignment actions will be templatized on type, I think, since RHS expressions will have had explicit conversion operations applied to force them to the type of the LHS.  So instances of the most generic expression type (`Expr<SomeType>`) should be pretty rare in the CFG and probably isolated to places like the operands of `STOP` and `PAUSE` statements.

Reply:
I believe the refined type is encoded in the type of the 'v' member of the GenericExprWrapper, right? An assignment statement's type could then simply project the type of the LHS expression.

Reply:
What I propose is an general `Assignment` action that wraps a variant whose alternatives are all (LHS, RHS) pairs of identically typed expressions.